postgis/gserialized_supportfn.c

```diff
@@ -0,0 +1,254 @@
+#include "nodes/nodeFuncs.h"
+#include "nodes/makefuncs.h"
+#include "parser/parse_func.h"
+#include "utils/array.h"
+#include "utils/builtins.h"
+#include "utils/lsyscache.h"
+#include "utils/numeric.h"
+#include "utils/syscache.h"
+
+/* PostGIS */
+#include "liblwgeom.h"
+
+typedef struct
+{
+	char *fn_name;
+	int   strategy_number;
+	int   nargs;
+	int   expand_arg;
+} IndexableFunction;
+
+const IndexableFunction IndexableFunctions[] = {
+	{"st_intersects", RTOverlapStrategyNumber, 2, 0},
+	{"st_dwithin", RTOverlapStrategyNumber, 3, 3},
+	{"st_contains", RTContainsStrategyNumber, 2, 0},
+	{"st_within", RTContainedByStrategyNumber, 2, 0},
+	{"st_touches", RTOverlapStrategyNumber, 2, 0},
+	{"st_3dintersects", RTOverlapStrategyNumber, 2, 0},
+	{"st_containsproperly", RTContainsStrategyNumber, 2, 0},
+	{"st_coveredby", RTContainsStrategyNumber, 2, 0},
+	{"st_overlaps", RTOverlapStrategyNumber, 2, 0},
+	{"st_covers", RTContainsStrategyNumber, 2, 0},
+	{"st_crosses", RTOverlapStrategyNumber, 2, 0},
+	{"st_dfullywithin", RTOverlapStrategyNumber, 3, 3},
+	{"st_3dintersects", RTOverlapStrategyNumber, 3, 3},
+	{"st_3ddwithin", RTOverlapStrategyNumber, 3, 3},
+	{"st_3ddfullywithin", RTOverlapStrategyNumber, 3, 3},
+	{"st_linecrossingdirection", RTOverlapStrategyNumber, 2, 0},
+	{NULL, 0, 0, 0}
+};
+
+static bool
+needsSpatialIndex(Oid funcid, IndexableFunction *idxfn)
+{
+	const IndexableFunction *idxfns = IndexableFunctions;
+	const char *fn_name = get_func_name(funcid);
+
+	do
+	{
+		if(strcmp(idxfns->fn_name, fn_name) == 0)
+		{
+			*idxfn = *idxfns;
+			return true;
+		}
+		idxfns++;
+	}
+	while (idxfns->fn_name);
+
+	return false;
+}
+
+static Oid
+opFamilyAmOid(Oid opfamilyoid)
+{
+	Form_pg_opfamily familyform;
+	// char *opfamilyname;
+	Oid opfamilyam;
+	HeapTuple familytup = SearchSysCache1(OPFAMILYOID, ObjectIdGetDatum(opfamilyoid));
+	if (!HeapTupleIsValid(familytup))
+		elog(ERROR, "cache lookup failed for operator family %u", opfamilyoid);
+	familyform = (Form_pg_opfamily) GETSTRUCT(familytup);
+	opfamilyam = familyform->opfmethod;
+	// opfamilyname = NameStr(familyform->opfname);
+	// elog(NOTICE, "%s: found opfamily %s [%u]", __func__, opfamilyname, opfamilyam);
+	ReleaseSysCache(familytup);
+	return opfamilyam;
+}
+
+static Oid
+expandFunctionOid(Oid geo_datatype)
+{
+	List *expandfn_name = list_make1(makeString("st_expand"));
+	Oid radius_datatype = FLOAT8OID; /* Should always be FLOAT8OID */
+	Oid expandfn_args[2];
+	const bool noError = true;
+	Oid expandfn_oid;
+
+	expandfn_args[0] = geo_datatype;
+	expandfn_args[1] = radius_datatype;
+	expandfn_oid = LookupFuncName(expandfn_name, 2, expandfn_args, noError);
+	if (expandfn_oid == InvalidOid)
+		elog(ERROR, "%s: unable to lookup 'st_expand(Oid[%u], Oid[%u])'", __func__, geo_datatype, radius_datatype);
+	return expandfn_oid;
+}
+
+Datum postgis_index_supportfn(PG_FUNCTION_ARGS);
+PG_FUNCTION_INFO_V1(postgis_index_supportfn);
+Datum postgis_index_supportfn(PG_FUNCTION_ARGS)
```

Comment:
```suggestion
PGDLLEXPORT Datum postgis_index_supportfn(PG_FUNCTION_ARGS)
```

The function gets left out of my windows build without the PGDLLEXPORT clause

Reply:
I can add it, but boy I would love to know *why* it works, and why 90% of our functions work without it, but some have it. Any help? I see spgist support functions have it. Actually all other instances of it seem to trace back to spgist. But still: why only for that one symbol?

Reply:
I did an experiment of moving the 

Datum postgis_index_supportfn(PG_FUNCTION_ARGS);

declarator

to the top of the file right after the includes and that seems to work such that I don't need the PGDLLEXPORT later.

That might be why our others have worked because the header defs were at the top.  Seems very fragile I know.

Reply:
Weird.